### DISCUSSION TITLE
Enforce strings for index keys

### DISCUSSION THREAD

**Discussion by jpike88** [rome/tools#4356] [Suggestions]
Index types should be locked to strings because of the unpredictable behaviors associated with using numbers as keys on an object.

If someone wants to use numbers in that way they should use the Map object which is designed for that.

Take this example showing how a problem can arise:
```
const x: { [k: number]: string } = {};

x[3] = 'test';

const keys = Object.keys(x); // this will have type of string[] so it's wrong
```

**Comment by ematipico** [comment]
Is there already a rule for that? What kind of bugs can this rule prevent?

**Reply by jpike88** [reply]
This snippet might be clearer showing how the use of numbers as keys can be a recipe for problems:

`Object.keys({ 1: "a", 2: "b", 3: "c" }).map(key => typeof key)`

returns

`["string", "string", "string"]`

I found an instance of this in my codebase which is quite large. It just seems like an anti-pattern to me, not much benefit in allowing it.

**Reply by ematipico** [reply]
That's how it works: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys

But if you iterate the object using a for..of loop, you'll get the correct types.

**Reply by jpike88** [reply]
ill close it as it's a deeper design problem than a linting problem
